### PR TITLE
test: add e2e framework

### PR DESCRIPTION
**Pull request by helayoty**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Add the e2e framework that will be needed to build all `scheduler-plugins` e2e tests cases.

#### Special notes for your reviewer:
CC: @Huang-Wei

### CODE REVIEW

**Comment by Huang-Wei** [comment]
> Add the 2e2 framework that will be needed to build all `scheduler-plugins` e2e tests cases.

@helayoty does this mean this PR is a preparation PR to set up the skeleton bits of e2e test suite, but not aimed to test any particular plugin functionality?

Some other questions:
- can we provide another option to test the single scheduler? (i.e., replacing the embed scheduler with the scheduler-plugins binary)
- do you have plans to cover e2e testing for controllers?

**Comment by helayoty** [comment]
> @helayoty does this mean this PR is a preparation PR to set up the skeleton bits of e2e test suite, but not aimed to test any particular plugin functionality?

Yes, I meant to have the base ready ASAP so other contributors can start jump in and help add more test cases.
> 
> Some other questions:
> 
> * can we provide another option to test the single scheduler? (i.e., replacing the embed scheduler with the scheduler-plugins binary)
I'm not sure I get the question. We didn't add tests cases for each plugin yet, but this is the plan. If you mean to not apply the helm chart with all plugins enabled, this is possible by using the same `helm` package from `e2e framework`.
> * do you have plans to cover e2e testing for controllers?
This will be included by default when testing the plugin, as you'll apply the CRD which calls the controller as part of the flow.

**Comment by Huang-Wei** [comment]
> I'm not sure I get the question. We didn't add tests cases for each plugin yet, but this is the plan. If you mean to not apply the helm chart with all plugins enabled, this is possible by using the same `helm` package from `e2e framework

Right now we provide two ways of installing the scheduler plugins: one is Helm chart, which is covered in this PR. While the other is to replace the scheduler binary, so my question is can we also support it in the e2e suite?

**Comment by helayoty** [comment]
>> I'm not sure I get the question. We didn't add tests cases for each plugin yet, but this is the plan. If you mean to not apply the helm chart with all plugins enabled, this is possible by using the same `helm` package from `e2e framework
> 
> Right now we provide two ways of installing the scheduler plugins: one is Helm chart, which is covered in this PR. While the other is to replace the scheduler binary, so my question is can we also support it in the e2e suite?

I believe yes we can. We can add this a task as well.

**Comment by Huang-Wei** [comment]
@helayoty to clarify, it seems this PR tries install helm chart from the prebuilt one?

```go
		if err := manager.RunInstall(helm.WithName("scheduler-plugins"),
			helm.WithChart(providerResourceAbsolutePath),
			helm.WithWait(), helm.WithTimeout("1m")); err != nil {
			klog.ErrorS(err, "failed to invoke helm install operation due to an error")
		}
```

Actually, to wire e2e to our CI pipeline, I'm expecting to build a temporary image/chart and use it in each PR's commit.

**Comment by helayoty** [comment]
> 


> @helayoty to clarify, it seems this PR tries install helm chart from the prebuilt one?
> 
> ```go
> 		if err := manager.RunInstall(helm.WithName("scheduler-plugins"),
> 			helm.WithChart(providerResourceAbsolutePath),
> 			helm.WithWait(), helm.WithTimeout("1m")); err != nil {
> 			klog.ErrorS(err, "failed to invoke helm install operation due to an error")
> 		}
> ```
> Actually, to wire e2e to our CI pipeline, I'm expecting to build a temporary image/chart and use it in each PR's commit.


For some reason, I forgot to add this part, thanks for the catch.

**Comment by helayoty** [comment]
@Huang-Wei kindly let me know how we can get this approved and merged.

**Comment by seanmalloy** [comment]
@helayoty the code change LGTM. Since we recently changed how we do vendoring in the project would you be able to rebase and run `go mod vendor`?

**Comment by helayoty** [comment]
> @helayoty the code change LGTM. Since we recently changed how we do vendoring in the project would you be able to rebase and run `go mod vendor`?

done

**Comment by helayoty** [comment]
@Huang-Wei @seanmalloy rebased and ready again.

**Review by Huang-Wei — COMMENTED** [review]
Sorry for the delay.

There maybe some glitch on loading the controller image into the kind node?

```
Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
LOCAL_REGISTRY=localhost:5000/scheduler-plugins  LOCAL_CONTROLLER_IMAGE=controller:latest LOCAL_IMAGE=kube-scheduler:latest go test -tags=e2e -v ./test/e2e
F0913 15:00:01.205968   45347 env.go:332] load docker image: kind: load docker-image failed: proc: error: exit status 1: Image: "localhost:5000/scheduler-plugins/controller:latest" with ID "sha256:8291681de217728a6aba80ebc514bb843215bc0748cc062a1344fdbde3dc99d8" not yet present on node "scheduler-plugins-e2e-test-control-plane", loading...
Image: "localhost:5000/scheduler-plugins/controller:latest" with ID "sha256:8291681de217728a6aba80ebc514bb843215bc0748cc062a1344fdbde3dc99d8" not yet present on node "scheduler-plugins-e2e-test-worker3", loading...
Image: "localhost:5000/scheduler-plugins/controller:latest" with ID "sha256:8291681de217728a6aba80ebc514bb843215bc0748cc062a1344fdbde3dc99d8" not yet present on node "scheduler-plugins-e2e-test-worker4", loading...
Image: "localhost:5000/scheduler-plugins/controller:latest" with ID "sha256:8291681de217728a6aba80ebc514bb843215bc0748cc062a1344fdbde3dc99d8" not yet present on node "scheduler-plugins-e2e-test-worker", loading...
Image: "localhost:5000/scheduler-plugins/controller:latest" with ID "sha256:8291681de217728a6aba80ebc514bb843215bc0748cc062a1344fdbde3dc99d8" not yet present on node "scheduler-plugins-e2e-test-worker2", loading...
ERROR: failed to load image: command "docker exec --privileged -i scheduler-plugins-e2e-test-worker2 ctr --namespace=k8s.io images import --digests --snapshotter=overlayfs -" failed with error: exit status 1

Command Output: unpacking localhost:5000/scheduler-plugins/controller:latest (sha256:e6f07d6cc88c40631ce159982e7188becd03f15450856de37b424f3eff5563ff)...ctr: content digest sha256:994393dc58e7931862558d06e46aa2bb17487044f670f310dffe1d24e4d1eec7: not found
goroutine 1 [running]:
k8s.io/klog/v2.stacks(0x1)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/klog/v2/klog.go:860 +0x90
k8s.io/klog/v2.(*loggingT).output(0x104824f00, 0x3, 0x0, 0x140000f0d90, 0x1, {0x10416eb88?, 0x104dac5b8?}, 0x10?, 0x0)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/klog/v2/klog.go:825 +0x614
k8s.io/klog/v2.(*loggingT).printDepth(0x104824f00, 0x3ef8308?, 0x0, {0x0, 0x0}, 0x0?, {0x1400007db40?, 0x1, 0x1})
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/klog/v2/klog.go:608 +0x174
k8s.io/klog/v2.(*loggingT).print(...)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/klog/v2/klog.go:590
k8s.io/klog/v2.Fatal(...)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/klog/v2/klog.go:1490
sigs.k8s.io/e2e-framework/pkg/env.(*testEnv).Run(0x140005a3300, 0x14000494f70?)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/sigs.k8s.io/e2e-framework/pkg/env/env.go:332 +0x458
sigs.k8s.io/scheduler-plugins/test/e2e.TestMain(0x14000080ec8?)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/test/e2e/main_test.go:49 +0x3b8
main.main()
	_testmain.go:47 +0x1d4

goroutine 24 [IO wait]:
internal/poll.runtime_pollWait(0x12be7f6d8, 0x72)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/netpoll.go:302 +0xa4
internal/poll.(*pollDesc).wait(0x140000f7080?, 0x14000800000?, 0x0)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/internal/poll/fd_poll_runtime.go:83 +0x2c
internal/poll.(*pollDesc).waitRead(...)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/internal/poll/fd_poll_runtime.go:88
internal/poll.(*FD).Read(0x140000f7080, {0x14000800000, 0x100c1, 0x100c1})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/internal/poll/fd_unix.go:167 +0x1e4
net.(*netFD).Read(0x140000f7080, {0x14000800000?, 0x1400003e898?, 0x8?})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/net/fd_posix.go:55 +0x2c
net.(*conn).Read(0x14000010008, {0x14000800000?, 0x1400003e898?, 0x8?})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/net/net.go:183 +0x38
crypto/tls.(*atLeastReader).Read(0x140005de108, {0x14000800000?, 0x0?, 0x102db79e8?})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/crypto/tls/conn.go:785 +0x44
bytes.(*Buffer).ReadFrom(0x1400003e978, {0x103ee3ad8, 0x140005de108})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/bytes/buffer.go:204 +0x94
crypto/tls.(*Conn).readFromUntil(0x1400003e700, {0x12bd940b8?, 0x14000010008}, 0xaa35?)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/crypto/tls/conn.go:807 +0xd8
crypto/tls.(*Conn).readRecordOrCCS(0x1400003e700, 0x0)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/crypto/tls/conn.go:614 +0xdc
crypto/tls.(*Conn).readRecord(...)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/crypto/tls/conn.go:582
crypto/tls.(*Conn).Read(0x1400003e700, {0x1400014d000, 0x1000, 0x1031f954c?})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/crypto/tls/conn.go:1285 +0x180
bufio.(*Reader).Read(0x14000115980, {0x14000130200, 0x9, 0x1031f8ed8?})
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/bufio/bufio.go:236 +0x1e4
io.ReadAtLeast({0x103ee3958, 0x14000115980}, {0x14000130200, 0x9, 0x9}, 0x9)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/io/io.go:331 +0xa4
io.ReadFull(...)
	/opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/io/io.go:350
golang.org/x/net/http2.readFrameHeader({0x14000130200?, 0x9?, 0x140d73eb2cf?}, {0x103ee3958?, 0x14000115980?})
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/golang.org/x/net/http2/frame.go:237 +0x5c
golang.org/x/net/http2.(*Framer).ReadFrame(0x140001301c0)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/golang.org/x/net/http2/frame.go:498 +0x84
golang.org/x/net/http2.(*clientConnReadLoop).run(0x140005bbf88)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/golang.org/x/net/http2/transport.go:2101 +0x100
golang.org/x/net/http2.(*ClientConn).readLoop(0x1400014a000)
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/golang.org/x/net/http2/transport.go:1997 +0x60
created by golang.org/x/net/http2.(*Transport).newClientConn
	/Users/weih/go/src/sigs.k8s.io/scheduler-plugins/vendor/golang.org/x/net/http2/transport.go:725 +0x924
FAIL	sigs.k8s.io/scheduler-plugins/test/e2e	144.893s
FAIL
make: *** [e2e-test] Error 1
```

File: test/e2e/main_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	providerResourceDirectory = "manifests/install/charts/as-a-second-scheduler/"
+	testenv                   env.Environment
+)
+
+func TestMain(m *testing.M) {
+	testenv = env.NewWithConfig(envconf.New())
+	// Create KinD Cluster
+	namespace := "scheduler-plugins"
+	testenv.Setup(
+		envfuncs.CreateKindClusterWithConfig(kindClusterName, "kindest/node:v1.22.2", "kind-config.yaml"),
```
[the remark this answers is not in the document]
Comment (helayoty) [reply test/e2e/main_test.go]: updated

**Pull request closed without merging**